beluga/include/beluga/sensor/data/regular_grid.hpp

```diff
@@ -74,8 +74,14 @@ class BaseRegularGrid2 : public ciabatta::ciabatta_top<Derived> {
    * \return Grid cell coordinates.
    */
   [[nodiscard]] Eigen::Vector2i cell_near(double x, double y) const {
-    const auto xi = static_cast<int>(std::floor(x / this->self().resolution()));
-    const auto yi = static_cast<int>(std::floor(y / this->self().resolution()));
+    const auto inv_resolution = 1. / this->self().resolution();
+    const auto scaled_x = x * inv_resolution;
+    const auto scaled_y = y * inv_resolution;
+    // Truncate towards zero. Use a poor man's replacement of
+    // `std::floor` because the later is awfully slow and this
+    // function is performance critical.
+    const auto xi = static_cast<int>(scaled_x) - (scaled_x < 0.0);
+    const auto yi = static_cast<int>(scaled_y) - (scaled_y < 0.0);
```

Comment:
Hmm, the following should make it right for all negative integers, at the expense of breaking it for all positive numbers closer than an epsilon to the immediately greater integer:
```suggestion
    const auto xi = static_cast<int>(scaled_x + std::numeric_limits<double>::epsilon()) - (scaled_x < 0.0);
    const auto yi = static_cast<int>(scaled_y + std::numeric_limits<double>::epsilon()) - (scaled_y < 0.0);
```

Reply:
can actually x and y be negative?
From a quick read to the code, it seems that they shouldn't be.

In that case, maybe we could try `std::trunc`?

Reply:
I can be negative, yes. It'll be discarded later because it falls outside of the grid.

I did not try `trunc` because most likely is as slow as `floor`. The problem with those functions is not with what they do, but in all the additional checks they need to do to be IEEE-754-compliant.

Reply:
`trunc` would also map the (-1.0, 0.0) range to 0, which we'd like to avoid.

Reply:
Maybe:

```cpp
static double opt_floor_impl(double x) {
  const auto int_x = static_cast<int>(x);
  bool is_not_integral = static_cast<double>(int_x) != x;
  return x - static_cast<double>(x < 0.0 && is_not_integral);
}
```

that should solve @hidmic comment and it's still maybe faster than the std version ....

Reply:
Given that in this case we only need the `floor`-like behavior to be correct in the ++ quadrant, I propose we leave the current version. 

Still, I think it's worth ticketing looking into other places where we may be using `std::floor` and familiy and determining if replacing those calls for a cheaper `floor`-like function. In this particular case, the function was the main cause of performance loss, and yet the functioin was invisible in the stack trace in perf. We may have further "performance leaks" in other places.

Reply:
> Given that in this case we only need the floor-like behavior to be correct in the ++ quadrant, I propose we leave the current version.

I see, makes sense.

Reply:
I created this ticket to track this: https://github.com/Ekumen-OS/beluga/issues/206